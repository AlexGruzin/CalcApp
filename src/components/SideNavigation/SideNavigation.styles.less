@import "~styles/generic.less";

.side-navigation {
  display: flex;
  flex-direction: column;
  max-width: 140px;
  min-width: 140px;
  background-color: var(--white);
}

.side-navigation__logo {
  margin: var(--space-x6) 0;
  height: 50px;
}

.side-navigation__link {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 140px;
  text-decoration: none;
}

.side-navigation__inner {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  width: 140px;
}

.side-navigation__content {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-left: 8px;
  width: 132px;
  margin: var(--space-x4) 0;
}

.side-navigation__icon {
  min-height: 30px;
  min-width: 30px;
}

.side-navigation__marker {
  display: flex;
  width: 8px;
  height: 100%;
}

.side-navigation__label {
  .montserrat-bold();
  margin: var(--space-x2) 0 0 0;
  font-size: var(--fontSize-superTiny);
}

.side-navigation__link.isActive .side-navigation__marker {
  background: var(--purple);
}

.side-navigation__link.link--settings::before {
  content: " ";
  display: flex;
  margin-bottom: var(--space-x4);
  margin-top: var(--space-x4);
  width: 100px;
  height: 1px;
  background: var(--grey200);
}
