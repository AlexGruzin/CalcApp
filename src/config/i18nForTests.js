import i18n from 'i18next';

// locales
import US from '../locales/en';

i18n.init({
  debug: false,
  fallbackLng: ['US'],
  lng: 'US',

  ns: ['auth', 'resetPassword'],
  defaultNS: 'auth',

  resources: {
    US,
  },
});

export default i18n;

