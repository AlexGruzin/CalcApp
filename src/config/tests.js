import React from 'react';
import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next'
import i18n from "./i18nForTests";
import jsdom from 'jsdom';
const { JSDOM } = jsdom;

/*
 * jest-dom
 * https://github.com/testing-library/jest-dom
 */
import '@testing-library/jest-dom/extend-expect';

/*
 * jest-fetch-mock
 * https://github.com/jefflau/jest-fetch-mock
 */
require('jest-fetch-mock').enableMocks();

if (global.document) {
  global.document.createRange = () => ({
    setStart: () => {
    },
    setEnd: () => {
    },
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
    createContextualFragment: str => JSDOM.fragment(str),
    // createContextualFragment: str => document.createElement('hr'),
  });
  global.open = () => undefined;
}

export const renderWithi18next = Component => {
  const Comp = React.cloneElement(Component, {
    changeLanguage: lng => {
      i18n.changeLanguage(lng);
      rerender(<I18nextProvider i18n={i18n}>{Comp}</I18nextProvider>);
    }
  });
  const defaultRender = render(
    <I18nextProvider i18n={i18n}>{Component}</I18nextProvider>
  );
  const { rerender } = defaultRender;
  return defaultRender;
};


// /**
//  * Wraps a component with an Intl Provider.
//  * @param {Object} component The component to render
//  */
// const withTestIntl = component => <IntlProvider locale="en">{component}</IntlProvider>;
//
// /*
//  * https://testing-library.com/docs/example-react-intl
//  */
// export const renderWithReactIntl = component => render(withTestIntl(component));

