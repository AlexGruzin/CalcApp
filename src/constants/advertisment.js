export const ADVERT = {
  CORDIAL: 'CORDIAL',
  FACEBOOK: 'FACEBOOK',
  GOOGLE_ADS: 'GOOGLE_ADS',
  GOOGlE_ANALYTICS: 'GOOGLE_ANALYTICS',
  INSTAGRAM: 'INSTAGRAM',
  KLAVIYO: 'KLAVIYO',
  TIKTOK: 'TIKTOK',
  TWITTER: 'TWITTER',
  SHOPIFY: 'SHOPIFY',
  SHOPIFY_SIGNUP: 'SHOPIFY_SIGNUP',
};

export const SOCIAL = {
  ENVELOPE: 'ENVELOPE',
  BITMAP: 'BITMAP',
};

export const ADVERT_ICONS = {
  [ADVERT.FACEBOOK]: 'facebook.png',
  [ADVERT.INSTAGRAM]: 'instagram.png',
  [SOCIAL.ENVELOPE]: 'envelope.png',
  [SOCIAL.BITMAP]: 'bitmap.png',
};
