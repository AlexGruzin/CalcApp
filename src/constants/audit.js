import { ADVERT } from 'constants/advertisment';

export const AUDIT_STORE_COMPLETE = 'STORE';
export const AUDIT_ANALYTICS_COMPLETE = 'ANALYTICS';
export const AUDIT_CHANNELS_COMPLETE = 'CHANNELS';

export const ADVERTISING_CHANNELS = [
  // {
  //   name: ADVERT.CORDIAL,
  //   image: 'cordial-logo.png',
  // },
  {
    name: ADVERT.GOOGLE_ADS,
    image: 'g-ads-logo.png',
  },
  // {
  //   name: ADVERT.FACEBOOK,
  //   image: 'facebook-logo.png',
  // },
  // {
  //   name: ADVERT.INSTAGRAM,
  //   image: 'instagram-logo.png',
  // },
  // {
  //   name: ADVERT.KLAVIYO,
  //   image: 'klaviyo-logo.png',
  // },
  // {
  //   name: ADVERT.TIKTOK,
  //   image: 'tiktok-logo.png',
  // },
  // {
  //   name: ADVERT.TWITTER,
  //   image: 'twitter-logo.png',
  // },
];

export const USER_STATUS = {
  FIRST_TIME: 'FIRST_TIME',
  AUDIT_PASSED: 'AUDIT_PASSED',
};
