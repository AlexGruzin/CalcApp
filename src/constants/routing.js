import { ROLES } from 'constants/roles';
import { USER_STATUS } from 'constants/audit';

export const HOME = '/';
export const LOG_IN = '/log_in';
export const SIGN_UP = '/sign_up';
export const DASHBOARD = '/dashboard';
export const RESET_PASSWORD = '/reset_password';
export const FORECAST = '/forecast';
export const CAMPAIGNS = '/campaigns';
export const SETTINGS = '/settings';
export const CONFIRM_OAUTH = '/confirm_oauth';
export const FORBIDDEN = '/forbidden';
export const CONFIRM_PASSWORD = '/confirmation_password';
export const CONFIRM_SIGN_UP = '/confirmation_registration';
export const CONFIRM_USER = '/confirmation_user';
export const CONFIRM_EMAIL_CHANGE = '/confirmation_email_change';
export const AUDIT = '/audit';
export const AUDIT_DASHBOARD = '/audit_dashboard';
export const SHOPIFY_SIGNUP = '/shopify_signup';
export const CONFIRM_SHOPIFY_SIGNUP = '/confirm_shopify_signup';
export const SURVEY = '/survey';
export const LOADER = '/loader';
export const PRIVACY_STATEMENT_LINK = '/privacy';
export const TERMS_LINK = '/terms';
export const WELCOME = '/welcome';

export const WELCOME_FIRST_TIME = `${WELCOME}?status=${USER_STATUS.FIRST_TIME}`;
export const WELCOME_AUDIT_PASSED = `${WELCOME}?status=${USER_STATUS.AUDIT_PASSED}`;

export const ROUTES_FULLY_ACCESSIBLE = [
  LOG_IN,
  HOME,
  SIGN_UP,
  CONFIRM_OAUTH,
  RESET_PASSWORD,
  FORBIDDEN,
  CONFIRM_PASSWORD,
  CONFIRM_USER,
  CONFIRM_SIGN_UP,
  SHOPIFY_SIGNUP,
  CONFIRM_SHOPIFY_SIGNUP,
  TERMS_LINK,
  PRIVACY_STATEMENT_LINK,
];

export const IFRAME_BLOCKED_ROUTES = [SIGN_UP];

export const ROUTES_PERMISSIONS = {
  [SETTINGS]: {
    [ROLES.USER]: true,
  },
  [FORECAST]: {
    [ROLES.USER]: true,
  },
  [CAMPAIGNS]: {
    [ROLES.USER]: true,
  },
  [DASHBOARD]: {
    [ROLES.USER]: true,
  },
  [AUDIT]: {
    [ROLES.USER]: true,
  },
  [AUDIT_DASHBOARD]: {
    [ROLES.USER]: true,
  },
  [CONFIRM_SIGN_UP]: {
    [ROLES.USER]: true,
  },
  [CONFIRM_USER]: {
    [ROLES.USER]: true,
  },
  [CONFIRM_PASSWORD]: {
    [ROLES.USER]: true,
  },
  [CONFIRM_EMAIL_CHANGE]: {
    [ROLES.USER]: true,
  },
  [SHOPIFY_SIGNUP]: {
    [ROLES.USER]: true,
  },
  [CONFIRM_SHOPIFY_SIGNUP]: {
    [ROLES.USER]: true,
  },
  [SURVEY]: {
    [ROLES.USER]: true,
  },
  [LOADER]: {
    [ROLES.USER]: true,
  },
  [WELCOME]: {
    [ROLES.USER]: true,
  },
};
