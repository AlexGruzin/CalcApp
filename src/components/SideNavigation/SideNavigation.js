import React from 'react';
import { NAVIGATION_OPTIONS, AUDIT_NAVIGATION_OPTIONS } from 'constants/dashboard';
import * as Routing from 'constants/routing';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';
import classNames from 'classnames';

import './SideNavigation.styles.less';

const SideNavigation = () => {
  const { t } = useTranslation();
  const selectedOption = window.location.pathname;
  let availableRoutes;
  switch (selectedOption) {
    case Routing.AUDIT: {
      availableRoutes = AUDIT_NAVIGATION_OPTIONS;
      break;
    }
    default: {
      availableRoutes = NAVIGATION_OPTIONS;
      break;
    }
  }

  return (
    <div className="side-navigation">
      <img className="side-navigation__logo" src="/svg/logo-purple.svg" alt="Calc logo" />
      {availableRoutes.map(option => {
        const isActive = option.route === selectedOption;
        const routeUniqueClass = `link--${option.route.slice(1)}`;
        return (
          <Link
            className={classNames('side-navigation__link', { isActive }, routeUniqueClass)}
            key={option.route}
            to={isActive ? '#' : option.route} // also pointer-event
            role="button"
          >
            <div className="side-navigation__inner">
              <div className="side-navigation__content">
                <img className="side-navigation__icon" src={`/svg/${option.icon}.svg`} alt="" />
                <h6 className="side-navigation__label">{t(option.label)}</h6>
              </div>
              <div className="side-navigation__marker" />
            </div>
          </Link>
        );
      })}
    </div>
  );
};

SideNavigation.propTypes = {};

SideNavigation.defaultProps = {};

export default SideNavigation;
