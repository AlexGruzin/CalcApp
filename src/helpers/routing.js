import { ROUTES_PERMISSIONS, ROUTES_FULLY_ACCESSIBLE, IFRAME_BLOCKED_ROUTES } from 'constants/routing';
import { ROLES } from 'constants/roles';

export const isInIframe = () => window.top !== window.self;

export const openNewTab = url => {
  const win = window.open(url, '_blank');
  if (win) {
    win.focus();
  }
};

export const redirectTopFrame = href => {
  window.top.location.href = href;
};

export const closeThisTab = () => {
  window.opener = null;
  window.open('', '_self');
  window.close();
};

export const checkAvailabilityPathByRoles = (targetPath, roles) => {
  /* order is necessary */
  if (isInIframe() && IFRAME_BLOCKED_ROUTES.includes(targetPath)) return { error: {} }; // in iframe some routes not allowed
  if (ROUTES_FULLY_ACCESSIBLE.includes(targetPath)) return { success: targetPath }; // these paths are accessible by anyone
  if (roles.includes(ROLES.ADMIN)) return { success: targetPath }; // admin got full access
  if (roles.length === 0) return { error: {} };

  const allowedRoutesForUser = roles.map(role => ROUTES_PERMISSIONS[targetPath][role]);
  const isAllowed = allowedRoutesForUser.some(item => item === true);
  return isAllowed ? { success: targetPath } : { error: {} };
};
