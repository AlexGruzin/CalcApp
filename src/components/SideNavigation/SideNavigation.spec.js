import React from 'react';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { renderWithi18next } from 'config/tests';
import { BrowserRouter } from 'react-router-dom';
import * as Routing from 'constants/routing';

import SideNavigation from './SideNavigation';

describe('SideNavigation tests', () => {
  it('Check snapshots of SideNavigation are equal.', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <SideNavigation />
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Check SideNavigation do redirect.', async () => {
    const { getAllByRole } = render(
      <BrowserRouter>
        <SideNavigation />
      </BrowserRouter>
    );
    const links = getAllByRole('button');

    await fireEvent.click(links[1]); // forecast

    await waitFor(() => {
      expect(window.location.pathname).toEqual(Routing.FORECAST);
    });
  });
});
