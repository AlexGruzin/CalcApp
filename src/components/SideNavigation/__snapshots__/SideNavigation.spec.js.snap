// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavigation tests Check snapshots of SideNavigation are equal. 1`] = `
<DocumentFragment>
  <div
    class="side-navigation"
  >
    <img
      alt="Calc logo"
      class="side-navigation__logo"
      src="/svg/logo-purple.svg"
    />
    <a
      class="side-navigation__link link--dashboard"
      href="/dashboard"
      role="button"
    >
      <div
        class="side-navigation__inner"
      >
        <div
          class="side-navigation__content"
        >
          <img
            alt=""
            class="side-navigation__icon"
            src="/svg/dashboard-icon.svg"
          />
          <h6
            class="side-navigation__label"
          >
            app:nav.dashboard
          </h6>
        </div>
        <div
          class="side-navigation__marker"
        />
      </div>
    </a>
    <a
      class="side-navigation__link link--forecast"
      href="/forecast"
      role="button"
    >
      <div
        class="side-navigation__inner"
      >
        <div
          class="side-navigation__content"
        >
          <img
            alt=""
            class="side-navigation__icon"
            src="/svg/forecast-icon.svg"
          />
          <h6
            class="side-navigation__label"
          >
            app:nav.forecast
          </h6>
        </div>
        <div
          class="side-navigation__marker"
        />
      </div>
    </a>
    <a
      class="side-navigation__link link--audit_dashboard"
      href="/audit_dashboard"
      role="button"
    >
      <div
        class="side-navigation__inner"
      >
        <div
          class="side-navigation__content"
        >
          <img
            alt=""
            class="side-navigation__icon"
            src="/svg/audit-icon.svg"
          />
          <h6
            class="side-navigation__label"
          >
            app:nav.audit
          </h6>
        </div>
        <div
          class="side-navigation__marker"
        />
      </div>
    </a>
    <a
      class="side-navigation__link link--campaigns"
      href="/campaigns"
      role="button"
    >
      <div
        class="side-navigation__inner"
      >
        <div
          class="side-navigation__content"
        >
          <img
            alt=""
            class="side-navigation__icon"
            src="/svg/campaigns-icon.svg"
          />
          <h6
            class="side-navigation__label"
          >
            app:nav.campaigns
          </h6>
        </div>
        <div
          class="side-navigation__marker"
        />
      </div>
    </a>
    <a
      class="side-navigation__link link--settings"
      href="/settings"
      role="button"
    >
      <div
        class="side-navigation__inner"
      >
        <div
          class="side-navigation__content"
        >
          <img
            alt=""
            class="side-navigation__icon"
            src="/svg/settings-icon.svg"
          />
          <h6
            class="side-navigation__label"
          >
            app:nav.settings
          </h6>
        </div>
        <div
          class="side-navigation__marker"
        />
      </div>
    </a>
  </div>
</DocumentFragment>
`;
